gateway-service/src/main/java/org/zowe/apiml/gateway/security/service/token/OIDCTokenProvider.java

```diff
@@ -10,72 +10,114 @@
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.message.BasicHeader;
+import org.apache.http.message.BasicNameValuePair;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
 import org.springframework.beans.factory.annotation.Value;
-import org.springframework.http.HttpEntity;
 import org.springframework.http.HttpHeaders;
-import org.springframework.http.HttpMethod;
-import org.springframework.http.ResponseEntity;
+import org.springframework.http.MediaType;
 import org.springframework.stereotype.Service;
-import org.springframework.web.client.RestClientException;
-import org.springframework.web.client.RestTemplate;
-import org.zowe.apiml.gateway.security.service.schema.OIDCAuthException;
 import org.zowe.apiml.security.common.token.OIDCProvider;
 
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
 import java.util.Base64;
+import java.util.List;
 
 @RequiredArgsConstructor
 @Service
 @Slf4j
 public class OIDCTokenProvider implements OIDCProvider {
 
-    @Value("${apiml.security.oAuth.clientId:}")
+    @Value("${apiml.security.oidc.clientId:}")
     private String clientId;
 
-    @Value("${apiml.security.oAuth.clientSecret:}")
+    @Value("${apiml.security.oidc.clientSecret:}")
     private String clientSecret;
 
-    @Value("${apiml.security.oAuth.validationUrl:}")
-    private String validationUrl;
-
-    @Value("${apiml.security.oAuth.enabled:false}")
+    @Value("${apiml.security.oidc.enabled:false}")
     private boolean isEnabled;
 
-    private final RestTemplate restTemplate;
+    @Autowired
+    @Qualifier("secureHttpClientWithoutKeystore")
+    private final CloseableHttpClient httpClient;
+
+
+    private static final ObjectMapper mapper = new ObjectMapper();
 
     @Override
-    public boolean isValid(String token) {
-        if (token == null || token.isEmpty() || !isEnabled) {
+    public boolean isValid(String token, String issuer) {
+        OIDCTokenClaims claims = introspect(token, issuer);
+        if (claims != null) {
+            return claims.getActive();
+        }
+        return false;
+    }
+
+    private OIDCTokenClaims introspect(String token, String issuer) {
+        if (StringUtils.isBlank(token) || !isEnabled) {
             log.debug("Either you did not enable the OIDC auth or you did not provide a valid token.");
-            throw new OIDCAuthException("A failure occurred when validating.");
+            return null;
+        }
+        if (StringUtils.isBlank(issuer) || !isValidURL(issuer)) {
+            log.warn("The OIDC token does not contain issuer claim or it is not valid URI. Cannot proceed with validation.");
+            return null;
         }
+        HttpPost post = new HttpPost(issuer + "/v1/introspect");
```

Comment:
It looks that keycloak is using just /introspect (according to RFC). The Ping Identity us using strange /as/introspect.oauth2 endpoint but it could be configured according to documentation. So currently, it will work only with OKTA out of the box. Do you have any suggestion here?

Reply:
Introspection endpoint is now configurable.